a73xq: camera: refactor xml, fix device specific issues

nicodotgit opened the pull request:
Refactored the XML for better readability and making working on mods for other devices much easier, based on UN1CA standards. Reorganized XML structure with section comments and proper indentation.

a73xq specific changes:
- Remove Log Video recording.
- Fix crash in Single Take mode.
- Switch back to standard Portrait mode.
- ~Re-add cutout animation.~

nicodotgit committed:
a73xq: camera: refactor xml, fix device specific issues

Refactored the XML for better readability and making working on mods for other devices much easier, based on UN1CA standards. Reorganized XML structure with section comments and proper indentation.
-
a73xq specific changes:
- Remove Log Video recording.
- Fix crash in Single Take mode.
- Switch back to standard Portrait mode.
- Re-add cutout animation.
---

Co-authored-by: utkustnr <utkuustuner@proton.me>

nicodotgit commented on target/a73xq/camera/camera-feature.xml:
```diff
@@ -11,334 +28,546 @@
-    <local name="BACK_CAMERA_RESOLUTION_4BY3_LARGE"      value="4000x3000"/>
-    <local name="BACK_CAMERA_RESOLUTION_FULL_RATIO"      value="4000x1800"/>
+    <local name="BACK_CAMERA_PICTURE_DEFAULT_RESOLUTION"    value="4000x3000"/>
+    <local name="BACK_CAMERA_RESOLUTION_16BY9_LARGE"        value="4000x2252"/>
```
This is to match the resolution that Samsung set in the stock XML of the A73.

nicodotgit commented on target/a73xq/camera/camera-feature.xml:
```diff
@@ -11,334 +28,546 @@
+    <local name="SUPPORT_CAMERA_ASSISTANT_CROP_ZOOM_X10"                value="false"/>
+    <local name="SUPPORT_CAMERA_ASSISTANT_DIGITAL_ZOOM_UPSCALE"         value="true"/>
+    <local name="SUPPORT_CAMERA_ASSISTANT_DOF_ADAPTER"                  value="false"/>
+	<local name="SUPPORT_CAMERA_ASSISTANT_ZOOM_X100"                    value="false"/>
```
Bruh.

nicodotgit commented on target/a73xq/camera/camera-feature.xml:
```diff
@@ -11,334 +28,546 @@
+    <local name="BOTTOM_GUIDE_LINE"                                                 value="0.681f"/>
+    <local name="CAMERA_QUICK_LAUNCH_USING_POWER_KEY"                               value="true"/>
+    <local name="DELAY_TIME_TO_RESUME_WHEN_SWITCH_FACING_WHILE_RECORDING"           value="100"/>
+    <local name="FLASH_OVERHEAT_LIMITATION_TEMP"                                    value="490"/>
```
Safe value, taken from the latest S25's XML.

nicodotgit commented on target/a73xq/camera/camera-feature.xml:
```diff
@@ -11,334 +28,546 @@
+    <local name="SUPPORT_SINGLE_TAKE_EXTENDED_CAPTURE"  value="true"/>
+    <local name="SUPPORT_SINGLE_TAKE_FILTERED_VIDEOS"   value="true"/>
+    <local name="SUPPORT_SINGLE_TAKE_HIGHLIGHT_VIDEOS"  value="true"/>
+    <local name="SUPPORT_SINGLE_TAKE_MULTI_CAMERA"      value="false"/>
```
This is the change that disables lens switching in Single Take and fixes the crash issue.

nicodotgit commented on target/a73xq/camera/camera-feature.xml:
```diff
@@ -11,334 +28,546 @@
+
+    <!-- ZOOM FEATURES -->
+
+    <local name="DEFAULT_MAX_ZOOM_VALUE"        value="1000"/>
```
There's no noticeable difference. I kept the real shortcut values in A73.

nicodotgit commented on target/a73xq/camera/camera-feature.xml:
```diff
@@ -11,334 +28,546 @@
+    <local name="BACK_CAMERA_RECORDING_DEFAULT_RESOLUTION"  value="1920x1080_AUTO_FPS"/>
+    <local name="BACK_CAMERA_RECORDING_MIN_RESOLUTION"      value="1280x720"/>
+    <!-- 20:9 Ratio -->
+    <local name="BACK_CAMCORDER_RESOLUTION_FEATURE_MAP_2400X1080"           value="true"    hdr="true"  hdr10="false"   hlg="false"     snapshot-support="true"     snapshot-size="4000x1868"   vdis="true"     super-vdis="true"   effect="true"   object-tracking="true"  seamless-zoom-support="true"    physical-zoom-supported="false"     external-storage-support="true"                         support ...
```
4000x1868 is the value that Samsung set for the A73. I didn't update the res map, but it doesn't affect functionality.

nicodotgit commented on target/a73xq/camera/camera-feature.xml:
```diff
@@ -11,334 +28,546 @@
+    <local name="BACK_CAMCORDER_RESOLUTION_FEATURE_MAP_3840X2160"           value="true"    hdr="true"  hdr10="false"   hlg="false"     snapshot-support="true"     snapshot-size="4000x2252"   vdis="true"     super-vdis="false"  effect="true"   object-tracking="true"  seamless-zoom-support="true"    physical-zoom-supported="false"     external-storage-support="true"                         support ...
+    <local name="BACK_CAMCORDER_RESOLUTION_FEATURE_MAP_3840X2160_24FPS"     value="true"    hdr="true"  hdr10="false"   hlg="false"     snapshot-support="true"     snapshot-size="4000x2252"   vdis="false"    super-vdis="false"  effect="true"   object-tracking="true"  seamless-zoom-support="true"    physical-zoom-supported="false"     external-storage-support="true"                         support ...
+    <local name="BACK_CAMCORDER_RESOLUTION_FEATURE_MAP_1920X1080"           value="true"    hdr="true"  hdr10="false"   hlg="false"     snapshot-support="true"     snapshot-size="4000x2252"   vdis="true"     super-vdis="true"   effect="true"   object-tracking="true"  seamless-zoom-support="true"    physical-zoom-supported="false"     external-storage-support="true"                         support ...
+    <local name="BACK_CAMCORDER_RESOLUTION_FEATURE_MAP_1920X1080_24FPS"     value="true"    hdr="true"  hdr10="false"   hlg="false"     snapshot-support="true"     snapshot-size="4000x2252"   vdis="false"    super-vdis="false"  effect="true"   object-tracking="true"  seamless-zoom-support="true"    physical-zoom-supported="false"     external-storage-support="true"                         support ...
+    <local name="BACK_CAMCORDER_RESOLUTION_FEATURE_MAP_1920X1080_60FPS"     value="true"    hdr="true"  hdr10="false"   hlg="false"     snapshot-support="true"     snapshot-size="4000x2252"   vdis="false"    super-vdis="false"  effect="false"  object-tracking="true"  seamless-zoom-support="true"    physical-zoom-supported="false"     external-storage-support="true"     uw-support="false"  support ...
+    <local name="BACK_CAMCORDER_RESOLUTION_FEATURE_MAP_1920X1080_AUTO_FPS"  value="true"    hdr="true"  hdr10="false"   hlg="false"     snapshot-support="true"     snapshot-size="4000x2252"   vdis="false"    super-vdis="false"  effect="false"  object-tracking="true"  seamless-zoom-support="true"    physical-zoom-supported="false"     external-storage-support="true"     uw-support="true"   support ...
+    <local name="BACK_CAMCORDER_RESOLUTION_FEATURE_MAP_1280X720"            value="true"    hdr="true"  hdr10="false"   hlg="false"     snapshot-support="true"     snapshot-size="4000x2252"   vdis="true"     super-vdis="true"   effect="true"   object-tracking="true"  seamless-zoom-support="true"    physical-zoom-supported="false"     external-storage-support="true"                         support ...
```
Intended. Didn't update the res map tho.

nicodotgit commented on target/a73xq/camera/camera-feature.xml:
```diff
@@ -11,334 +28,546 @@
+    <local name="SUPPORT_CAMERA_ASSISTANT_AUTO_LENS_SWITCHING"          value="true"/>
+    <local name="SUPPORT_CAMERA_ASSISTANT_CROP_ZOOM_X2"                 value="false"/>
+    <local name="SUPPORT_CAMERA_ASSISTANT_CROP_ZOOM_X10"                value="false"/>
+    <local name="SUPPORT_CAMERA_ASSISTANT_DIGITAL_ZOOM_UPSCALE"         value="true"/>
```
All the Camera Assistant functions I disabled were not working.

nicodotgit commented on target/a73xq/camera/camera-feature.xml:
```diff
@@ -11,334 +28,546 @@
+    -->
+
+    <!-- Main Modes -->
+    <local name="SHOOTING_MODE_FUN"                 back="FUN"                  front="FUN"                 enable="true"   more="false"    order="1"/>
```
Even if it's useless, this mode is enabled on A73 by default.

nicodotgit commented on target/a73xq/camera/camera-feature.xml:
```diff
@@ -11,334 +28,546 @@
+    <local name="CAMERA_QUICK_LAUNCH_USING_POWER_KEY"                               value="true"/>
+    <local name="DELAY_TIME_TO_RESUME_WHEN_SWITCH_FACING_WHILE_RECORDING"           value="100"/>
+    <local name="FLASH_OVERHEAT_LIMITATION_TEMP"                                    value="490"/>
+    <local name="MAX_BURST_CAPTURE_COUNT"                                           value="1500"/>
```
In tests I conducted myself, the camera crashed after approximately 1910 burst images. In my opinion, 1500 is a safe value.

nicodotgit commented on target/a73xq/camera/camera-feature.xml:
```diff
@@ -11,334 +28,546 @@
+    <local name="SUPPORT_LIVE_BEAUTY_IN_AUTO_MODE"      value="true"/>
+    <local name="SUPPORT_LIVE_BEAUTY_MANUAL_RESHAPE"    value="true"/>
+    <local name="SUPPORT_LIVE_BEAUTY_SHAPE_CORRECTION"  value="true"/>
+    <local name="SUPPORT_SELFIE_TONE_MODE"              value="false"/>
```
This wasn't working, so it was disabled.

nicodotgit commented on target/a73xq/camera/camera-feature.xml:
```diff
@@ -11,334 +28,546 @@
+    <!-- CAMERA ASSISTANT FEATURES -->
+
+    <local name="SUPPORT_CAMERA_ASSISTANT"                              value="true"/>
+    <local name="SUPPORT_CAMERA_ASSISTANT_ADAPTIVE_PIXEL"               value="true"/>
```
https://github.com/salvogiangri/UN1CA/pull/614#discussion_r2654719289

nicodotgit commented on target/a73xq/camera/camera-feature.xml:
```diff
@@ -11,334 +28,546 @@
+    <local name="SUPPORT_BACK_WIDE_PRO"                         value="true"/>
+    <local name="SUPPORT_HISTOGRAM"                             value="true"/>
+    <local name="SUPPORT_PRO_AE_PRIORITY_MODE"                  value="true"/>
+    <local name="SUPPORT_PRO_RESET"                             value="true"/>
```
The reset function worked before anyway; I added it to have a more robust XML file.

Fede2782 commented:
What is this? I know my opinion doesn't count much, but for me this is a no.

 a73 stock camera-feature wouldn't normally require any change as it got official One UI 8, so all modes which are present there would work

Samsung camera feature has a standard format, rewriting it completely is a crazy nonsense idea.

Is a 800 lines change PR really required to fix something which would already work? 

I'd wait for the maintainer @utkustnr and also why he didn't use stock camera feature in first place.

salvogiangri commented:
> I'd wait for the maintainer @utkustnr and also why he didn't use stock camera feature in first place.

A73 3.x builds already ship a modded camera-feature xml. This PR probably aims on updating such config to enhance it/fix bugs

salvogiangri commented:
> * Re-add cutout animation.

This has been removed from the One UI 8.0 camera app.

nicodotgit commented:
> > * Re-add cutout animation.
> 
> This has been removed from the One UI 8.0 camera app.

I had no idea about this, but having it there doesn't affect its functionality anyway.

To clarify, the XML keeps everything that @utkustnr tuned for our device; I've only improved the format, which now is much closer to the standard format, and modified the things I already mentioned in the main comment. I do find spaces and tabs unnecessary, but they're there to maintain the XML format that Utku followed and that's currently in the repo since https://github.com/salvogiangri/UN1CA/commit/278644b75cf61882e84a2941c89ee9bcda4b3bd2.

The comments in the XML might be unnecessary for A73, but they're useful for someone who would like to modify the XML for other devices, a base to start from and not waste hours figuring out what each thing in camera-feature is about and what could cause problems.

Fede2782 commented:
> > > * Re-add cutout animation.
> > 
> > 
> > This has been removed from the One UI 8.0 camera app.
> 
> I had no idea about this, but having it there doesn't affect its functionality anyway.
> 
> To clarify, the XML keeps everything that @utkustnr tuned for our device; I've only improved the format, which now is much closer to the standard format, and modified the things I already mentioned in the main comment. I do find spaces and tabs unnecessary, but they're there to maintain the XML format that Utku followed and that's currently in the repo since [278644b](https://github.com/salvogiangri/UN1CA/commit/278644b75cf61882e84a2941c89ee9bcda4b3bd2).
> 
> The comments in the XML might be unnecessary for A73, but they're useful for someone who would like to modify the XML for other devices, a base to start from and not waste hours figuring out what each thing in camera-feature is about and what could cause problems.

Having a reference XML is a great idea, I'm not hiding that. However I'd keep it separated from the shipped A73 one. Also the Camera Assistant options are redundant, if not present it is assumed to be false already. For example, A series only define the main camera assistant clause.

nicodotgit commented:
> > > > * Re-add cutout animation.
> > > 
> > > 
> > > This has been removed from the One UI 8.0 camera app.
> > 
> > 
> > I had no idea about this, but having it there doesn't affect its functionality anyway.
> > To clarify, the XML keeps everything that @utkustnr tuned for our device; I've only improved the format, which now is much closer to the standard format, and modified the things I already mentioned in the main comment. I do find spaces and tabs unnecessary, but they're there to maintain the XML format that Utku followed and that's currently in the repo since [278644b](https://github.com/salvogiangri/UN1CA/commit/278644b75cf61882e84a2941c89ee9bcda4b3bd2).
> > The comments in the XML might be unnecessary for A73, but they're useful for someone who would like to modify the XML for other devices, a base to start from and not waste hours figuring out what each thing in camera-feature is about and what could cause problems.
> 
> Having a reference XML is a great idea, I'm not hiding that. However I'd keep it separated from the shipped A73 one. Also the Camera Assistant options are redundant, if not present it is assumed to be false already. For example, A series only define the main camera assistant clause.

This is the place I thought was good to leave that more polished base for others. I honestly forgot to comment the Camera Assistant false features; I was mainly focused on the structure in general and those values were one of the last things I touched and focused on.

If as a team you consider it's better to keep what you already have, I can close this PR and just solve the problems that currently exist in A73 specifically (fix Single Take multi-lens crash, switch back to standard Portrait mode, disable Log video recording due to incompatibility) without touching anything else.

Fede2782 commented:
> > > > > * Re-add cutout animation.
> > > > 
> > > > 
> > > > This has been removed from the One UI 8.0 camera app.
> > > 
> > > 
> > > I had no idea about this, but having it there doesn't affect its functionality anyway.
> > > To clarify, the XML keeps everything that @utkustnr tuned for our device; I've only improved the format, which now is much closer to the standard format, and modified the things I already mentioned in the main comment. I do find spaces and tabs unnecessary, but they're there to maintain the XML format that Utku followed and that's currently in the repo since [278644b](https://github.com/salvogiangri/UN1CA/commit/278644b75cf61882e84a2941c89ee9bcda4b3bd2).
> > > The comments in the XML might be unnecessary for A73, but they're useful for someone who would like to modify the XML for other devices, a base to start from and not waste hours figuring out what each thing in camera-feature is about and what could cause problems.
> > 
> > 
> > Having a reference XML is a great idea, I'm not hiding that. However I'd keep it separated from the shipped A73 one. Also the Camera Assistant options are redundant, if not present it is assumed to be false already. For example, A series only define the main camera assistant clause.
> 
> This is the place I thought was good to leave that more polished base for others. I honestly forgot to comment the Camera Assistant false features; I was mainly focused on the structure in general and those values were one of the last things I touched and focused on.
> 
> If as a team you consider it's better to keep what you already have, I can close this PR and just solve the problems that currently exist in A73 specifically (fix Single Take multi-lens crash, switch back to standard Portrait mode, disable Log video recording due to incompatibility) without touching anything else.

It's a great idea to have a reference somewherre. I mean, I'd just edit the camera assistant ones. Also the 3d recording for galaxy XR is unrelated to the STEREO config. Some A series have galaxy XR record without having the STEREO config. Anyways, happy new year.

nicodotgit commented:
> > > > > > * Re-add cutout animation.
> > > > > 
> > > > > 
> > > > > This has been removed from the One UI 8.0 camera app.
> > > > 
> > > > 
> > > > I had no idea about this, but having it there doesn't affect its functionality anyway.
> > > > To clarify, the XML keeps everything that @utkustnr tuned for our device; I've only improved the format, which now is much closer to the standard format, and modified the things I already mentioned in the main comment. I do find spaces and tabs unnecessary, but they're there to maintain the XML format that Utku followed and that's currently in the repo since [278644b](https://github.com/salvogiangri/UN1CA/commit/278644b75cf61882e84a2941c89ee9bcda4b3bd2).
> > > > The comments in the XML might be unnecessary for A73, but they're useful for someone who would like to modify the XML for other devices, a base to start from and not waste hours figuring out what each thing in camera-feature is about and what could cause problems.
> > > 
> > > 
> > > Having a reference XML is a great idea, I'm not hiding that. However I'd keep it separated from the shipped A73 one. Also the Camera Assistant options are redundant, if not present it is assumed to be false already. For example, A series only define the main camera assistant clause.
> > 
> > 
> > This is the place I thought was good to leave that more polished base for others. I honestly forgot to comment the Camera Assistant false features; I was mainly focused on the structure in general and those values were one of the last things I touched and focused on.
> > If as a team you consider it's better to keep what you already have, I can close this PR and just solve the problems that currently exist in A73 specifically (fix Single Take multi-lens crash, switch back to standard Portrait mode, disable Log video recording due to incompatibility) without touching anything else.
> 
> It's a great idea to have a reference somewherre. I mean, I'd just edit the camera assistant ones. Also the 3d recording for galaxy XR is unrelated to the STEREO config. Some A series have galaxy XR record without having the STEREO config. Anyways, happy new year.

Happy New Year to you too! 🎉
My comments on the XML about STEREO were based on what happened in my tests with this specific device. Do you think you could elaborate on what you mentioned? For example, on what device does this happen, and if it's not as I described in the comment, what would this feature actually be for? If I'm wrong about something, it would be good to know so it can be improved or removed.
I'll check up on the Camera Assistant features, thanks for mentioning it.

Fede2782 commented:
gts9fe has 3D audio recording without having the STEREO config. I unfortunately don't have the device or the firmware with me so can't provide the xml as reference. I recall there was also a library related to this but, again, that has revealed to be false as gts9fe doesn't have it.

nicodotgit committed:
a73xq: camera: correct contents in camera-feature.xml

- Update camera resolution values in Resolution Map.
- Remove display cutout animation and Samsung XR mentions.
- Move the Camera Assistant features that were set to "false" to
  the corresponding comment.

Suggested-by: @salvogiangri
Suggested-by: @Fede2782
